compiler/semantic/analyzer.go

```diff
@@ -49,26 +58,26 @@ func newAnalyzer(ctx context.Context, source *data.Source, head *lakeparse.Commi
 	}
 }
 
-func (a *analyzer) enterScope() {
-	a.scope = NewScope(a.scope)
-}
-
-func (a *analyzer) exitScope() {
-	a.scope = a.scope.parent
-}
-
-func (a *analyzer) buildFrom(op dag.Op) (dag.Op, error) {
-	switch op := op.(type) {
+func HasSource(s dag.Seq) bool {
+	switch op := s[0].(type) {
 	case *dag.FileScan, *dag.HTTPScan, *dag.PoolScan, *dag.LakeMetaScan, *dag.PoolMetaScan, *dag.CommitMetaScan, *dag.DeleteScan:
-		return nil, nil
+		return true
 	case *dag.Fork:
-		return a.buildFrom(op.Paths[0][0])
+		return HasSource(op.Paths[0])
 	case *dag.Scope:
-		return a.buildFrom(op.Body[0])
+		return HasSource(op.Body)
+	}
+	return false
+}
+
+func (a *analyzer) addDefaultSource(s *dag.Seq) error {
```

Comment:
Nit: We call a dag.Seq parameter `seq` everywhere else.
```suggestion
func (a *analyzer) addDefaultSource(s *dag.Seq) error {
```